File: taskrun/src/main/java/com/goexp/galgame/data/task/others/GetTrueCVTask.java
package com.goexp.galgame.data.task.others;

import com.goexp.common.util.Strings;
import com.goexp.galgame.common.model.CV;
import com.goexp.galgame.data.db.importor.mongdb.GameDB;
import com.goexp.galgame.data.db.query.mongdb.CVQuery;
import com.goexp.galgame.data.db.query.mongdb.GameQuery;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.stream.Collectors;

public class GetTrueCVTask {

    private static CVQuery cvQuery = new CVQuery();
    private static GameQuery gameQuery = new GameQuery();
    private static GameDB gameDB = new GameDB();


    public static void main(String[] args) {

        final Logger logger = LoggerFactory.getLogger(GetTrueCVTask.class);


        var cvMap = CV.getMap(cvQuery.cvList());


        logger.info("Init OK");


        gameQuery.list()
//                .stream()
                .parallelStream()
                .peek(game -> {

                    if (game.gameCharacterList != null)
                        game.gameCharacterList = game.gameCharacterList.stream()
                                .peek(gameCharacter -> {
                                    if (!Strings.isEmpty(gameCharacter.cv)) {
                                        var cv = cvMap.get(gameCharacter.cv.trim().toLowerCase());

                                        if (cv != null) {
                                            gameCharacter.trueCV = cv.name;
//
                                            logger.debug("CV:{},trueCV:{}", gameCharacter.cv, gameCharacter.trueCV);
                                        }
                                    }
                                })
                                .collect(Collectors.toUnmodifiableList());
                })
                .forEach(game -> {
                    gameDB.updateChar(game);
                });
    }
}

File: task/src/main/scala/com/goexp/galgame/data/script/source/getchu/local/GetTrueCVTask.scala
package com.goexp.galgame.data.script.source.getchu.local

import com.goexp.common.util.string.Strings
import com.goexp.galgame.common.model.CV
import com.goexp.galgame.common.model.game.GameCharacter
import com.goexp.galgame.data.source.getchu.importor.GameDB
import com.goexp.galgame.data.source.getchu.query.{CVQuery, GameFullWithCharQuery}
import com.mongodb.client.model.Filters.{not, eq => same}
import com.typesafe.scalalogging.Logger

import scala.jdk.CollectionConverters._


object GetTrueCVTask {
  private val logger = Logger(GetTrueCVTask.getClass)

  type Person = GameCharacter

  def getMap(cvList: List[CV]) =
    cvList.to(LazyList)
      .flatMap(cv => {
        cv.nameStr.split("[=＝]")
          .map((name: String) => {
            val c = new CV
            c.name = cv.name
            c.trueName = name.trim.toLowerCase
            c
          })
      })
      .map(cv => cv.trueName -> cv)
      .toMap


  def main(args: Array[String]): Unit = {

    val list = CVQuery().scalaList().toList
    val localCV = getMap(list)

    logger.info("Init OK")

    GameFullWithCharQuery()
      .where(not(same("gamechar", null)))
      .scalaList().to(LazyList)
      .filter(g => Option(g.gameCharacters).map(_.size).getOrElse(0) > 0)
      .map(g => {

        var change = false

        g.gameCharacters = g.gameCharacters.asScala
          .map(p => {
            var tempP = p

            def isTarget(p: Person) = Strings.isNotEmpty(p.cv) && Strings.isEmpty(p.trueCV)

            val cv = tempP.cv.trim.toLowerCase
            if (isTarget(tempP) && localCV.contains(cv)) {
              val trueCV = localCV(cv)
              tempP = tempP.copy(trueCV = trueCV.name)

              logger.info(s"CV:${tempP.cv},trueCV:${tempP.trueCV}  Game: ${g.name} ")
              change = true
            }

            tempP
          }).asJava

        (change, g)

      })
      .foreach {
        case (true, game) =>
          GameDB.updateChar(game)
        case _ =>
      }
  }


}